vms/platformvm/network/gossip.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package network
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/network/p2p"
+	"github.com/ava-labs/avalanchego/network/p2p/gossip"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/vms/platformvm/txs"
+	"github.com/ava-labs/avalanchego/vms/platformvm/txs/mempool"
+)
+
+var (
+	_ p2p.Handler                = (*txGossipHandler)(nil)
+	_ gossip.Marshaller[*txs.Tx] = (*txMarshaller)(nil)
+	_ gossip.Gossipable          = (*txs.Tx)(nil)
+)
+
+// txGossipHandler is the handler called when serving gossip messages
+type txGossipHandler struct {
+	p2p.NoOpHandler
+	appGossipHandler  p2p.Handler
+	appRequestHandler p2p.Handler
+}
+
+func (t txGossipHandler) AppGossip(
+	ctx context.Context,
+	nodeID ids.NodeID,
+	gossipBytes []byte,
+) {
+	t.appGossipHandler.AppGossip(ctx, nodeID, gossipBytes)
+}
+
+func (t txGossipHandler) AppRequest(
+	ctx context.Context,
+	nodeID ids.NodeID,
+	deadline time.Time,
+	requestBytes []byte,
+) ([]byte, error) {
+	return t.appRequestHandler.AppRequest(ctx, nodeID, deadline, requestBytes)
+}
+
+type txMarshaller struct{}
+
+func (t txMarshaller) MarshalGossip(tx *txs.Tx) ([]byte, error) {
+	return tx.Bytes(), nil
+}
+
+func (t txMarshaller) UnmarshalGossip(bytes []byte) (*txs.Tx, error) {
+	parsed, err := txs.Parse(txs.Codec, bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return parsed, nil
+}
+
+func newGossipMempool(
+	mempool mempool.Mempool,
+	log logging.Logger,
+	txVerifier TxVerifier,
+	maxExpectedElements uint64,
+	falsePositiveProbability float64,
+	maxFalsePositiveProbability float64,
+) (*gossipMempool, error) {
+	bloom, err := gossip.NewBloomFilter(maxExpectedElements, falsePositiveProbability)
+	return &gossipMempool{
+		Mempool:                     mempool,
+		log:                         log,
+		txVerifier:                  txVerifier,
+		bloom:                       bloom,
+		maxFalsePositiveProbability: maxFalsePositiveProbability,
+	}, err
+}
+
+type gossipMempool struct {
+	mempool.Mempool
+	log                         logging.Logger
+	txVerifier                  TxVerifier
+	emptyBlockPermitted         bool
+	maxFalsePositiveProbability float64
+
+	lock  sync.RWMutex
+	bloom *gossip.BloomFilter
+}
+
+func (g *gossipMempool) Add(tx *txs.Tx) error {
+	txID := tx.ID()
+	if _, ok := g.Mempool.Get(txID); ok {
+		return fmt.Errorf("tx %s dropped: %w", tx.ID(), mempool.ErrDuplicateTx)
+	}
+
+	if reason := g.Mempool.GetDropReason(txID); reason != nil {
+		// If the tx is being dropped - just ignore it
+		//
+		// TODO: Should we allow re-verification of the transaction even if it
+		// failed previously?
+		return reason
```

Comment:
NB: this is a behavior change. Previously you could re-issue a tx and it would clear the dropped status. I only did this because the X-Chain also does it and we were trying to keep these implementations similar.